code/modules/mob/living/carbon/xenomorph/castes/hivemind/hivemind.dm

```diff
@@ -191,11 +214,80 @@
 	. = ..()
 	if(isnull(parent))
 		return
+
+	if(!CHECK_BITFIELD(datum_flags, DF_ISPROCESSING))
+		START_PROCESSING(SSslowprocess, src)
+
 	var/health_percent = round((max_integrity / obj_integrity) * 100)
 	switch(health_percent)
 		if(-INFINITY to 25)
-			to_chat(parent, "<span class='xenohighdanger'>Your core is under attack, and dangerous low on health!</span>")
+			to_chat(parent, "<span class='xenohighdanger'>Your core is under attack, and is dangerously low on health!</span>")
 		if(26 to 75)
-			to_chat(parent, "<span class='xenodanger'>Your core is under attack, and low on health!</span>")
+			to_chat(parent, "<span class='xenodanger'>Your core is under attack, and is low on health!</span>")
 		if(76 to INFINITY)
 			to_chat(parent, "<span class='xenodanger'>Your core is under attack!</span>")
+
+
+/mob/living/carbon/xenomorph/hivemind/upgrade_xeno(newlevel) //where we also upgrade the max health of our core and restore it to maximum health
+	. = ..()
+	to_chat(src, "<span class='xenonotice'>Our core strengthens, and becomes more resilient!</span>")
+	update_core()
+	core.obj_integrity = core.max_integrity
```

Comment:
I've no big qualms about this, but should it not just increase the max, and immediately start processing the heal?

Reply:
The precedent is you get maxed out on upgrading/evolving, so it should probably apply for your core as well.